### DISCUSSION TITLE
audiobooks: better UI design

### DISCUSSION THREAD

**Discussion by llemeurfr** [edrlab/thorium-reader#1749] [Ideas]
We should make the design better looking. 

The best design I've seen recently is on the [Jellybooks web app ](https://www.jellybooks.com/cloud_reader/audiobooks/little_women) (a pure JS implementation), with a responsive design, a larger cover, clean SVG buttons and progress bar.

Another [implementation by De Marque](https://player.cantookaudio.com/aHR0cHM6Ly9yZWFkaXVtLm9yZy93ZWJwdWItbWFuaWZlc3QvZXhhbXBsZXMvRmxhdGxhbmQvbWFuaWZlc3QuanNvbg==)

Should we ask a frontend developer some help? 

Original issue: https://github.com/readium/readium-desktop/issues/956

Related issues:
#955
#955
#942

Follow-up to PR #938

**Comment by danielweck** [comment]
Not that the minimum width of Thorium's window is quite wide, in fact is it impossible to reach portrait aspect ratio when resizing the window (i.e. no "mobile" style responsive design).

EDIT: well, I personally use an external display in vertical orientation, so strictly-speaking Thorium's window can in fact reach a portrait aspect ratio, but I meant that in practice, this is an unlikely occurrence :)

**Comment by danielweck** [comment]
To illustrate my comment above, here is Thorium's mininum window dimensions (I am not able to resize any smaller):

![Screenshot 2020-03-23 at 19 37 34](https://user-images.githubusercontent.com/625201/77356206-c85a1180-6d3d-11ea-9e2d-4a56a7ac00e4.png)

**Comment by danielweck** [comment]
...and, here is Thorium on my external vertical display :)

![Screenshot 2020-03-23 at 19 38 50](https://user-images.githubusercontent.com/625201/77356323-f9d2dd00-6d3d-11ea-8d44-66b38441f4c5.png)

**Comment by danielweck** [comment]
...so, I am not convinced (yet) about the "responsive" argument in the case of Thorium (for the "mobile" / portrait layout), unless we drastically change how window dimensions are currently restricted / dictated by the publication info dialog, right-side slide-out menus, etc.

**Comment by danielweck** [comment]
This is Thorium in wide landscape mode ("fullscreen" on my HD laptop display):

![Screenshot 2020-03-23 at 19 41 17](https://user-images.githubusercontent.com/625201/77356484-4d452b00-6d3e-11ea-8a22-3c8e2c244d07.png)

In this case, we can definitely afford to increase the dimensions of the cover image, and placing the audio controls on the right side of the cover image (instead of below) may indeed be a improvement.

**Comment by danielweck** [comment]
As for the audio controls (buttons): I have my personal opinion about the "look and feel", but regardless of our convergence / divergence about how "nice" the UI is, we can definitely discuss how to improve the implementation, which right now is located in the navigator component, which does not benefit from Thorium's localized strings (accessibility labels, even if not displayed as text in the UI), and the configurable keyboard shortcuts to activate commands related to audiobook playback.

Therefore, in a future development iteration we will need to address this significant shortcoming by designing a UI paradigm (in Thorium) that overlaps with the central area where the navigator component normally renders its own content (but remember: actual audio playback is still handled by the navigator, therefore requiring the usual sandboxed webview runtime with hidden HTML5 audio element, event listeners, etc.).

![Screenshot 2020-03-23 at 19 45 51](https://user-images.githubusercontent.com/625201/77356810-eecc7c80-6d3e-11ea-9ac2-7d1fb2074017.png)

**Comment by danielweck** [comment]
There was already an issue logged for this, so I am deleting is as duplicate:
https://github.com/readium/readium-desktop/issues/956

**Comment by danielweck** [comment]
Another thought: the `r2-testapp-js` "test app" will probably never implement its own audio controls UI (unnecessary development costs, just to test the navigator component). Therefore, `r2-navigator-js` will continue to provide its UI as a "sensible default", but will allow host apps (like Thorium) to deactivate the default UI. In concrete terms, the host app will register a set of event listeners in order to interact with the audio playback progress events (there are quite a few!), in order to trigger updates in its own UI (e.g. progress bar, play/pause state, etc.)

**Comment by danielweck** [comment]
The current "design" is built with basic CSS shapes, quite angular. Much like the BBC "Gel" icons: https://www.bbc.co.uk/gel/guidelines/iconography#audio-and-visual

Conversely, the Material Design "rounded" icons are softer and more elegant:
https://material.io/resources/icons/?style=round

(scroll down to find the play/pause/rewind/etc. icons)

**Comment by danielweck** [comment]
Updated `r2-navigator-js` player style (SVG icons), and added speed/rate control:

![Screenshot 2020-04-07 at 12 18 05](https://user-images.githubusercontent.com/625201/78663533-310ec580-78ca-11ea-95ba-2a406ead69f6.png)

![Screenshot 2020-04-07 at 12 18 29](https://user-images.githubusercontent.com/625201/78663541-353ae300-78ca-11ea-8d0f-0fe4c881508c.png)

![Screenshot 2020-04-07 at 12 18 17](https://user-images.githubusercontent.com/625201/78663549-3835d380-78ca-11ea-85ac-755772abdd77.png)

**Comment by danielweck** [comment]
Note that the screenshots above show the minimal reader window size (constrained width/height dimensions). Also note that the time indicators underneath the audio player's own timeline (left and right-aligned) show by default the current time and the total track time, respectively. However the labels are clickable and can toggle between remaining time (negative value) and percentage progression, respectively.

Also note that the bottom bar and its accompanying left/right arrow buttons (located on the right hand side) has not been tweaked for audiobooks yet (other than the "more information" popup which shows progress data). The bottom progress bar encompasses the entire publication reading order (i.e. multiple audio tracks), so it is not redundant per-se. However it looks redundant in the user interface, which is quite crowded. To be discussed further.

**Comment by danielweck** [comment]
bug triage: this clearly needs to be moved to a "discussion", this is not an actionable issue (plus, we decided that responsive design is not on the roadmap, due to Thorium's minimum window width)
